File: src/lib/tool-manifest/tools/select-named-region.ts
import { z } from 'zod';
import { useAiSession } from '@/hooks/useImageContext';
import { useEditorStore } from '@/store';
import { maskStore } from '@/core/mask-store';
import { ackSchema } from '../shared-schemas';
import type { ToolManifest } from '../types';

const input = z.object({
  label: z.string().min(1).describe('Region or Object label.'),
  commit: z.boolean().default(true).describe('If true (default), the mask is committed immediately (selection is locked). If false, it stays in active/preview state.'),
});

export const selectNamedRegionTool: ToolManifest<typeof input, typeof ackSchema> = {
  name: 'select_named_region',
  kind: 'mutate',
  description:
    'Arm a named region/Object as the active selection. Prefers a committed Object with this label (instant); falls back to an AI-proposed region (creates a candidate). Always prefer this over raw coordinate-based segmentation when a name covers the goal.',
  inputSchema: input,
  outputSchema: ackSchema,
  handler: ({ label, commit }) => {
    // 1. Try committed Object by label.
    const matchingObject = maskStore
      .all()
      .find((m) => m.label && m.label.toLowerCase() === label.toLowerCase());
    if (matchingObject) {
      useEditorStore.getState().setActiveObjectId(matchingObject.id);
      return { ok: true, message: `Selected Object "${matchingObject.label}".` };
    }
    // 2. Fall back to AI-precomputed region.
    const ctx = useAiSession.getState().context;
    const region = ctx?.candidateRegions?.find((r) => r.label.toLowerCase() === label.toLowerCase());
    if (!region) {
      return { ok: false, message: `No region or Object with label "${label}". Call list_named_regions to see what is available.` };
    }
    if (!region.maskRef) {
      return { ok: false, message: `Region "${label}" exists but has no mask. The image may have been analysed without segmentation.` };
    }
    useEditorStore.getState().setActiveMask(region.maskRef);
    if (commit) useEditorStore.getState().commitMask();
    return { ok: true, message: `Selected region "${label}" (${commit ? 'committed' : 'preview'}).` };
  },
};

File: backend/app/tools/atomic/select_named_region.py
from __future__ import annotations

from pydantic import BaseModel, Field

from app.state.document import SessionDocument
from app.tools.base import BackendTool, ToolPermissions


class _UnknownRegion(KeyError):
    """Mapped to unknown_region in the envelope by the registry."""


class _ScopeUnresolvable(KeyError):
    """Mapped to scope_unresolvable."""


class _Input(BaseModel):
    label: str = Field(min_length=1)
    commit: bool = True


class _Output(BaseModel):
    ok: bool
    state: str  # "active" | "committed"
    mask_id: str


class SelectNamedRegionTool(BackendTool[_Input, _Output]):
    name = "select_named_region"
    kind = "mutate"
    description = (
        "Arm a Claude-named region as the active selection. Prefer this over raw "
        "coordinate-based segmentation when a named region covers the goal."
    )
    input_schema = _Input
    output_schema = _Output
    permissions = ToolPermissions(requires_image=False)

    async def handler(self, doc: SessionDocument, input: _Input) -> _Output:  # noqa: A002
        ctx = doc.get_image_context("in-default")
        if ctx is None:
            raise _ScopeUnresolvable("no image context yet")
        region = next((r for r in ctx.candidate_regions if r.label == input.label), None)
        if region is None:
            raise _UnknownRegion(f"no region named {input.label!r}")
        mask = next((m for m in doc.masks.values() if m.label == input.label), None)
        if mask is None:
            raise _ScopeUnresolvable(
                f"region {input.label!r} has no registered mask; "
                "call select_by_point or seed via /analyze pre-segmentation"
            )
        if input.commit:
            doc.active_mask_id = None
            doc.committed_mask_id = mask.id
            state = "committed"
        else:
            doc.active_mask_id = mask.id
            state = "active"
        doc.emit_selection_changed(mask.id, state, input.label)
        return _Output(ok=True, state=state, mask_id=mask.id)
